tpm_quote/src/command.rs

```diff
@@ -0,0 +1,97 @@
+use crate::error::Error;
+
+use tss_esapi::{
+    handles::KeyHandle,
+    interface_types::{
+        algorithm::HashingAlgorithm, session_handles::AuthSession, structure_tags::AttestationType,
+    },
+    structures::{
+        Attest, AttestInfo, Data, HashScheme, PcrSelectionList, Signature, SignatureScheme,
+    },
+    Context,
+};
+
+use std::convert::TryFrom;
+
+pub(crate) fn attest_pcr(
+    context: &mut Context,
+    key_handle: &KeyHandle,
+    pcr_selection_list: PcrSelectionList,
+    nonce: Option<&[u8]>,
+) -> Result<(Attest, Signature), Error> {
+    let qualifying_data = if let Some(nonce) = nonce {
+        Data::try_from(nonce.to_vec()).map_err(|e| Error::AttestationError(format!("{e}")))
+    } else {
+        Data::try_from(vec![0xff; 16]).map_err(|e| Error::AttestationError(format!("{e}")))
+    }?;
+
+    let (attestation, signature) = context.execute_with_nullauth_session(|ctx| {
+        ctx.quote(
+            *key_handle,
+            qualifying_data,
+            SignatureScheme::EcDsa {
+                hash_scheme: HashScheme::new(HashingAlgorithm::Sha256),
+            },
+            pcr_selection_list.clone(),
+        )
+    })?;
+
+    if attestation.attestation_type() != AttestationType::Quote {
+        return Err(Error::TpmError(
+            "Attestation type of the returned value is not indicating Quote".to_owned(),
+        ));
+    }
+
+    match attestation.attested() {
+        AttestInfo::Quote { info } => {
+            if info.pcr_digest().len() == 0 {
+                return Err(Error::QuoteError("Digest in QuoteInfo is empty".to_owned()));
+            }
+
+            if info.pcr_selection() != &pcr_selection_list {
+                return Err(Error::QuoteError(
+                    "QuoteInfo selection list did not match the input selection list".to_owned(),
+                ));
+            }
+        }
+        _ => {
+            return Err(Error::QuoteError(
+                "Attested did not contain the expected variant.".to_owned(),
+            ));
+        }
+    }
+
+    Ok((attestation, signature))
+}
+
+#[allow(dead_code)]
+pub(crate) fn certify_key_handle(
+    context: &mut Context,
+    sign_key_handle: &KeyHandle,
+    key_handle: &KeyHandle,
+    nonce: Option<&[u8]>,
+) -> Result<(Attest, Signature), Error> {
+    let qualifying_data = if let Some(nonce) = nonce {
+        nonce.to_vec()
+    } else {
+        vec![0xff; 16]
+    };
+
+    let (attestation, signature) = context.execute_with_sessions(
+        (
+            Some(AuthSession::Password),
+            Some(AuthSession::Password),
+            None,
+        ),
+        |ctx| {
+            ctx.certify(
+                (*key_handle).into(),
+                *sign_key_handle,
+                Data::try_from(qualifying_data).unwrap(),
```

Comment:
Do not unwrap?

Reply:
See 4c89b1e036c2e9819789901d7ffb13bb0212a650.